pkg/checks/external/main.go

```diff
@@ -407,123 +407,96 @@ func (ext *Checker) watchForCheckerPodShutdown(shutdownEventNotifyC chan struct{
-		if err == nil {
-			ext.log("created a pod watcher successfully")
-			return watcher
-		}
-
-		// if we have failed, up our fail count and exit it its been too many
-		fails++
-		if fails > maxFails {
-			ext.log("reached maximum fails for starting a watcher. triggering fatal shutdown.")
-			log.Fatal("Unable to start watch for checker pod shutdown:", err)
-		}
-
-		ext.log("error when watching for checker pod shutdown:", err.Error())
-		time.Sleep(time.Second) // wait between retries to start a watch
-	}
+	// start a new watch request
+	return podClient.Watch(listOptions)
 }
 
 // waitForDeletedEvent watches a channel of results from a pod watch and notifies the returned channel when a
 // removal is observed.  The supplied abort channel is for shutting down gracefully.
-func (ext *Checker) waitForDeletedEvent(eventsIn <-chan watch.Event, sawRemovalChan chan struct{}, stoppedChan chan struct{}) {
+func (ext *Checker) waitForDeletedEvent(w watch.Interface) chan error {
 
 	ext.wg.Add(1)
 	defer ext.wg.Done()
 
 	// restart the watcher repeatedly forever until we are told to shutdown
 	ext.log("starting pod shutdown watcher")
 
-	// watch events for a removal
-	for e := range eventsIn {
-		ext.log("got a result when watching for pod to remove")
-		switch e.Type {
-		case watch.Modified: // this section is entirely informational
-			ext.log("checker pod shutdown monitor saw a modified event.")
-			p, ok := e.Object.(*apiv1.Pod)
-			if !ok {
-				ext.log("checker pod shutdown monitor saw a modified event and the object was not a pod. skipped.")
-				continue
-			}
-			ext.log("checker pod shutdown monitor saw a modified event. the pod changed to ", p.Status.Phase)
-			return
-		case watch.Deleted:
-			ext.log("checker pod shutdown monitor saw a deleted event. notifying that pod has shutdown")
-			sawRemovalChan <- struct{}{}
-			return
-		case watch.Error:
-			ext.log("khcheck monitor saw an error event")
-			e, ok := e.Object.(*metav1.Status)
-			if ok {
-				ext.log("pod removal monitor had an error when watching for pod changes:", e.Reason)
+	// watch events for a removal and read until the watcher empties entirely, but only notify once
+	var haveNotified bool
+	var err error
+	outChan := make(chan error, 2)
+
+	go func() {
+		for e := range w.ResultChan() {
+			ext.log("got a result when watching for pod to remove")
+			switch e.Type {
+			case watch.Modified: // this section is entirely informational
+				ext.log("checker pod shutdown monitor saw a modified event.")
+				p, ok := e.Object.(*apiv1.Pod)
+				if !ok {
+					ext.log("checker pod shutdown monitor saw a modified event and the object was not a pod. skipped.")
+					break
+				}
+				ext.log("checker pod shutdown monitor saw a modified event. the pod changed to ", p.Status.Phase)
+			case watch.Deleted: // we saw a deleted event, so notify upstream, but only once
+				if !haveNotified {
+					outChan <- nil
+					haveNotified = true
+				}
```

Comment:
do we need to have a `haveNotified` or can we just break / return out of this forloop? It doesn't seem like we need to go through the rest of this forloop, until we reach a deleted event in which then we exit out of this function/watch?

Reply:
updated